### ISSUE TITLE
Start/configure plugins from the frontend?

### ISSUE THREAD

**Issue by adrientetar** [google/xi-editor#147]
I see that "xi-syntect-plugin" is currently hardcoded in run_plugins.

Should there instead be an RPC command by which the frontend asks backend to start a plugin, e.g. a "start" command with a path argument relative to the xi-core binary.

"start_plugin" path: "xi-syntect-plugin" -> plugin_id (-1 if failed to start?)

(on Android I rename binaries to .so to get them packaged in the apk so specifying path is extra helpful)

Also, there should be a way to e.g. tell the syntax highlighter to pick a particular language/theme. I assume it's up to syntect-plugin to implement relevant RPC commands? How would the backend pass-through? Would the command to the plugin be encapsulated in a command to the backend like:

"send_to_plugin" id: plugin_id command: JSONObject (maybe returns a result value?)

**Comment by adrientetar** [comment]
Ah, I just noticed the "debug_run_plugin" command in the code, nice. We should come up with the full-fledged plugin commands and document that.

**Comment by gnzlbg** [comment]
@raphlinus I would like to write a small plugin that implements common-lisp as an scripting language for xi (using SBCL as an interpreter), but I am having trouble finding out where to start (e.g. what does my plugin exactly need to implement, how do I set it up, etc.). Do we have an "example" plugin somewhere?

**Comment by adrientetar** [comment]
@gnzlbg Have you seen [syntect-plugin](https://github.com/google/xi-editor/tree/master/rust/syntect-plugin) in the tree?

Currently xi-core is hardwired to start that plugin if you send "debug_run_plugin" from the frontend.

**Comment by gnzlbg** [comment]
@adrientetar Yes, from there it just seems that I need to write a similar main function and just implement `caching_plugin::Handler` for my state object. But since syntect-plugin is a bit of "hardwired" I was wondering whether user plugins need to implement something else as well.

**Comment by adrientetar** [comment]
I think there's just no mechanism to start an arbitrary plugin right now (I just tweak run_plugin.rs by hand), which is why I posted this issue as a proposal.

**Comment by raphlinus** [comment]
Yes, I want to have a whole configuration mechanism for running plugins, as well as plumbing more information (file extension) so the plugin can adjust its behavior.

@gnzlbg: the best guide is currently looking at the syntect-plugin source. However, the plugin interface (like so much of xi) is not yet stable. The parts about notifying the plugin of changes to the document should be pretty solid.

I'm curious, what do you want to do?

**Comment by gnzlbg** [comment]
@raphlinus I wanted to hook an higher-level language as a plugin to script xi (I was thinking of SBCL).
